website/src/pages/docs/custom-codegen/document-transform.mdx

```diff
@@ -0,0 +1,117 @@
+import { Callout } from '@theguild/components'
+
+# Document Transform
+
+## Basic Usage
+
+Each plugin can also provide a function `transformDocuments` for transforming `documents`:
+
+```js {5-14}
+const { visit } = require('graphql')
+
+module.exports = {
+  plugin() {}, // Nothing to do
+  transformDocuments(schema, documents, config) {
+    return documents.map(documentFile => {
+      documentFile.document = visit(documentFile.document, {
+        Field(node) {
+          // This function triggered per each field
+        }
+      })
+      return documentFile
+    })
+  }
+}
+```
+
+You can specify the plugin with `documentTransforms` field in your config:
+
+```ts {9}
+import { CodegenConfig } from '@graphql-codegen/cli'
+
+const config: CodegenConfig = {
+  schema: 'http://localhost:4000/graphql',
+  documents: ['src/**/*.tsx'],
+  generates: {
+    './src/gql/': {
+      preset: 'client',
+      documentTransforms: ['./my-transform.js'],
+      plugins: []
+    }
+  }
+}
+
+export default config
+```
+
+<Callout>
+  If you specify it with only the `plugins` field and not with the `documentTransforms` field, the `transformDocuments`
+  function will not be executed.
+</Callout>
+
+## Example
+
+For example, let's remove a `@loacalOnlyDirective` directive from documents:
+
+```js
+const { visit, print } = require('graphql')
+
+module.exports = {
+  plugin(schema, documents, config) {
```

Comment:
I am not really sure if we need this hook/function. What benefit is it trying to solve?

could this file not simply be the followng?

```ts

module.exports = function transformDocuments (...args) => {
  
}

```

Reply:
In terms of the current implementation, I'm using [Plugin mechanism](https://github.com/dotansimha/graphql-code-generator/blob/f16fc4c7973a7d51689151539db363bd3d03faba/packages/graphql-codegen-cli/src/codegen.ts#L322) to execute`transformDocuments`. Currently, we get the `transformDocuments` function from Plugin.

Regarding your question, are you suggesting we create a plugin without the `plugin` function?
Currently, the `plugin` function is not optional, but I can change it to be optional. However, I have concerns that making the `plugin` function optional may confuse users and may not be a desirable change to the API. 🤔 

Another option would be not to use the Plugin mechanism at all. Instead, we could introduce a new concept called something like **DocumentTransform** that works independently of Plugin. Just as Preset is a separate concept from Plugin, DocumentTransform is separate from Plugin.
This would be a slightly over-engineered solution, and most implementations would probably be copies of Plugin, but it would make  API cleaner. :sparkles:
If this design is acceptable, I will implement it. 🚀

Reply:
Fixed https://github.com/dotansimha/graphql-code-generator/pull/8723#issuecomment-1435936804